### ISSUE TITLE
System.Diagnostics.Process doesn't allow to "wait and kill" without a race

### ISSUE THREAD

**Issue by Dmitry-Me** [dotnet/corefx#7343]
Suppose I have a `System.Diagnostics.Process` bound to some running process. I want to wait until that process exits and if it fails to exit in five seconds I want that process terminated. Here's a piece of pseudocode:

    if( !process.WaitForExit( 5000 ) ) {
        process.Kill();
    }
    process.Dispose();

The problem is the process might exit between `WaitForExit()` and `Kill()` and then this code gets a rather generic exception. There's no way to write the code above such that there's no potential for the race condition between the other process exiting and this process making the decision of whether to terminate the other process or not.

Same problem on desktop.

There're three approaches to how this can be resolved.

The first one is that `Kill()` does nothing if the process has exited. This would violate the contract stated in the documentation but it's actually the right thing to do. The purpose of `Kill()` is "get rid of the process" and if it so happens the process has exited - okay, nothing to do. I cannot imagine a reasonable case where someone would blindly call `Kill()` and then catch the possible exception and make any decisions on what to do based on whether there was an exception.

The second one is that a parameter with a default value is added to `Kill()` - such that it turns into `Kill( bool throwIfExited = true )`. This prevents violating the current contract but honestly it's stupid.

The third one is crafting some special exception exactly for this case - such that the caller can reliably detect that the process has exited by programmatically inspecting the exception. This would break the contract and it's also stupid.

**Comment by viktor-svub** [comment]
Hi! Just my 2c -- is throwing a specific exception derived from `InvalidOperationException` still a contract breach? I would _not_ consider it one, as any existing handlers will still catch it and technically it _is-an_  instance of the same class, isn't it? 
As for the other alternative -- defaults are inlined to caller, making them part of contract and a liability. 
I'd rather recommend a `bool TryKill()` or `bool KillIfExists()` non-throwing alternative.

**Comment by Dmitry-Me** [comment]
@viktor-svub Throwing a derived exception could become a breach if the caller compares the exception type for equality. Dunno how reasonable that would be, perhaps there're some scenarios where it makes sense. Anyway when your documentation says code throws `SomeException` and it actually throws `SomeDerivedException` that formally is a contract breach.

**Comment by viktor-svub** [comment]
@Dmitry-Me : Right you are, that makes sense. Then an alternate method with meaningful name is probably the best remaining alternative? Adding an argument to `Kill()` would be runtime-dependency breaking, as defaults are just syntactic sugar.

**Comment by Dmitry-Me** [comment]
@viktor-svub Yeap, I guess `KillIfExists()` would be just fine.

**Comment by Priya91** [comment]
@Dmitry-Me 
```
if( !process.WaitForExit( 5000 ) ) {
    try { process.Kill(); } catch (InvalidOperationException) { }
}
process.Dispose();
```

doesn't work? The exception thrown in case of Process exited is InvalidOperationException as documented.

**Comment by Dmitry-Me** [comment]
@Priya91 It sort of works but
* this doesn't look like a clear API - you have to expect a race to get motivated to write such code
* an exception introduces overhead for no good reason
* an exception can trigger a breakpoint when run under debugger for no good reason

**Comment by Priya91** [comment]
> an exception introduces overhead for no good reason
> an exception can trigger a breakpoint when run under debugger for no good reason

This justification applies to all exceptions that are caught and have no-op, doesn't mean it's not useful in other scenarios.

Modifying the existing API is not an option for compat reasons, and I fail to grasp the reason in introducing a new API. If you can come up with some data on why exception handling is bad for your scenario with some perf numbers or other usage scenarios, then please update the first issue comment with an api speclet like [here](https://github.com/dotnet/corefx/issues/271)

**Comment by Dmitry-Me** [comment]
@Priya91 Yes, and most of the time having an exception which is handled with a no-op is just useless. All you get is a breakpoint in debugger which you cannot do anything useful with anyway. However this part is minor. The major part is you have to first be aware of the potential race, and then craft this workaround instead of using a "get rid of this process" method. The current API displays has design, it's not hard to change it without breaking compatibility.

**Comment by Priya91** [comment]
@Dmitry-Me If we have a new API say ```bool TryKill()```, which return true if the process was killed with a kill signal and false if not. How will you know when the API returns false, the kill didn't happen because of some OS error, or trying to kill a remote process or if the process exited by itself before kill? If you get false, you are unsure if the process is alive (in case of first 2 reasons) or not (process exited by itself). That is, the state of process running is unknown in case of false from TryKill, and this is not adding any value over Kill which clearly denotes if the process is killed or not.

**Comment by Dmitry-Me** [comment]
@Priya91 Obviously such `TryKill()` is not of much use. It should be `KillIfStillRunning()` which can return `void` and just kills the process no matter if it exited or not and throws an exception if it cannot kill the process. It may also return `bool` to indicate whether the process has exited itself but I don't see much use in that.

**Comment by Priya91** [comment]
@Dmitry-Me I misunderstood the proposal, can you update the issue comment with an api speclet, u can use the example from [here](https://github.com/dotnet/corefx/issues/271). The speclet has to clearly define the purpose of the API through usage scenarios, and the behavior. Once we have the proposal we can push this through API review.

**Comment by jnm2** [comment]
On the derived exception sidetrail, I'm not convinced that it ever breaks a contract to throw a more derived exception any more than it breaks a contract to return a more derived value. If the contract is implemented following the LSP, it will never matter.

Reflection, whether involving exact instance types or members, is always likely to break if used improperly.
